reconcile/utils/saasherder/saasherder.py

```diff
@@ -1884,12 +1885,17 @@ def validate_promotions(self) -> bool:
         if not (self.state and self._promotion_state):
             raise Exception("state is not initialized")
 
+        has_errors = False
```

Comment:
save errors to `has_errors` and mutate in each iteration, with too many `continue` in nested loop is a little confusing. consider extract a method `validate_promotion(self, promotion) -> bool`, then you can do 

```python
return any(self.validate_promotion(promotion) for promotion in self.promotions)
```

if you don't want to early terminate validations, then collect results to list first

```python
        errors = [self.validate_promotion(promotion) for promotion in self.promotions]
        return any(errors)
```

Reply:
Would like to avoid bigger refactors in the validate function at this point - considering this now more of a styling issue.

Reply:
and if we let `validate_*` method return `True` means valid, `False` means invalid, then it's even more readable


```python
        return all(self.validate_promotion(promotion)
                   for promotion in self.promotions
                   if promotion is not None and promotion.subscribe)
```

and both `validate_promotions` and `validate_promotion` are easier to read and test.

Reply:
I hear you and I agree. But I would really love to get unblocked on this :bow: without adding much more stuff around saasherder. This is not a functional problem. Im happy to open a ticket to invest more time into saasherder complexity in general.

Reply:
> is it covered by any tests?
All saasherder tests are currently single publisher based, so this scenario is not covered.
Ive added multiple publisher scenario in the new tests, but they focus on soakDays. Will now see if I can add a generalized test for multi publisher.

Reply:
a new pr just touch this code block after current pr merged is good, no need for a big refactor, whenever adding new logic, should maintain complexity instead of letting it grow. similar to test coverage, if we have code quality report, the goal is not to make it worse for new changes.

Reply:
Test adjusted with multi-channel examples.